### Métodos Numéricos Computacionales
#### Prácticas computacionales de "Solución de sistemas de ecuaciones lineales"

Instrucciones para los Ejercicios

1. **Trabajo en Grupo:**
   - Los ejercicios deben ser resueltos y entregados en grupo.
   - La cantidad de integrantes por grupo será definida el día de la actividad, así como la fecha límite para la entrega.

2. **Uso de Google Colab y Compartir:**
   - Este notebook debe ser copiado al GitHub o Google Drive de alguno de los integrantes del grupo.
   - El grupo será responsable de programar las soluciones, realizar las pruebas y enviar el trabajo final al profesor.

3. **Implementación de los Ejercicios:**
   - Cada ejercicio debe ser implementado de manera que cumpla con los objetivos específicos descritos en cada problema.
   - El código debe devolver claramente la información calculada de acuerdo a lo solicitado.

4. **Calidad del Código:**
   - El código debe ejecutarse sin errores.
   - Es obligatorio incluir **comentarios explicativos** para describir las ideas y conceptos implícitos en el código, facilitando la comprensión de su lógica.

5. **Envío del Trabajo:**
   - Una vez completado, el notebook debe ser enviado a través de Moodle.
   - En caso de dudas, pueden contactarme por correo electrónico a **marcelo.danesi@utec.edu.uy**.

6. **Orientaciones Adicionales:**
   - Asegúrense de que todas las celdas de código hayan sido ejecutadas antes de enviar.
   - Incluyan el nombre completo y correo electrónico de todos los integrantes al inicio del notebook.
   - Si utilizan referencias externas, menciónenlas de forma adecuada.

¡Buena suerte y aprovechen la práctica para consolidar los conceptos de métodos numéricos!

#### **Métodos directos**

En la clase anterior, abordamos métodos algebraicos para la resolución de matrices, pero ¿cómo podemos trasladar este problema de encontrar la solución directa a Python? A continuación, veremos un ejemplo y algunos ejercicios que nos prepararán para la implementación de métodos iterativos, como el método de Jacobi y Gauss-Seidel.

##### **Ejemplo 1: Resolución de un sistema de ecuaciones lineales**


**Descripción del problema:**

Considere el sistema de ecuaciones lineales representado por la matriz

$$A = \left[
\begin{array}{rrr|r}
4 & -1 & 2 \\
0 & -3 & 1  \\
-1 & 2 & 5  
\end{array}
\right]$$

y el vector

$$b =\left[
\begin{array}{c}
15  \\
6  \\
11  
\end{array}
\right]$$

en la ecuación $Ax = b$. Nuestro objetivo es encontrar la solución exacta del sistema utilizando Python.

**Objetivo:**
- Implementar la resolución del sistema usando la función `numpy.linalg.solve`.
- Verificar la solución obtenida.

**Salidas:**
1. El vector solución $x$.


**Ejemplo de salida esperada:**

$\text{Solución}: [ 2. -1.  3.]$

In [ ]:
import numpy as np

# Definimos la matriz de coeficientes A
A = np.array([
    [4, -1, 2],
    [0, -3, 1],
    [-1, 2, 5]
], dtype=float)

# Definimos el vector de términos independientes b
b = np.array([15, 6, 11], dtype=float)

# Resolvemos el sistema de ecuaciones Ax = b
x = np.linalg.solve(A, b)

# Mostramos la solución obtenida
print("Solución:", x)

Solución: [ 2. -1.  3.]


##### **Ejemplo 2: Triangularización superior con `numpy.triu`**


**Descripción del problema:**

Utilice la función `numpy.triu` para obtener la forma triangular superior de la matriz

$$A = \left[
\begin{array}{rrr}
4 & -1 & 2 \\
0 & -3 & 1  \\
-1 & 2 & 5  
\end{array}
\right].$$


**Objetivo:**
- Extraer la matriz triangular superior.

**Salidas:**
1. Matriz triangular superior.


In [ ]:
import numpy as np

# Definimos la matriz de coeficientes A
A = np.array([
    [4, -1, 2],
    [0, -3, 1],
    [-1, 2, 5]
], dtype=float)

# Obtenemos la matriz triangular superior
A_triu = np.triu(A)

# Mostramos el resultado
print("Matriz triangular superior:")
print(A_triu)

Matriz triangular superior:
[[ 4. -1.  2.]
 [ 0. -3.  1.]
 [ 0.  0.  5.]]


Nos ejercicios a seguir, considere sempre a matriz

$$A = \left[
\begin{array}{rrr}
0 & 2 & -1 \\
1 & 1 & 4  \\
6 & -2 & 3  
\end{array}
\right].$$


##### **Ejercicio 1: Triangularización inferior con `numpy.tril`**

**Descripción del problema:**

Utilice la función `numpy.tril` para obtener la forma triangular inferior de la matriz dada $A$.

**Objetivo:**
- Extraer la matriz triangular inferior.

**Salidas:**
1. Matriz triangular inferior.

In [2]:
import numpy as np

# Definimos la matriz de coeficientes A
A = np.array([
    [0, 2, -1],
    [1, 1, 4],
    [6, -2, 3]
], dtype=float)

# Obtenemos la matriz triangular superior
A_tril = np.tril(A)

# Mostramos el resultado
print("Matriz triangular Inferior:")
print(A_tril)

Matriz triangular Inferior:
[[ 0.  0.  0.]
 [ 1.  1.  0.]
 [ 6. -2.  3.]]


##### **Ejercicio 2: Factorización LU con `scipy.linalg.lu`**

**Descripción del problema:**

Utilice `scipy.linalg.lu` para descomponer la matriz fija $A$ en sus componentes $L$ y $U$. En caso de dudas sobre el uso del comando, consulte la documentación oficial de la biblioteca SciPy.

**Objetivo:**
- Descomponer la matriz en factores triangulares inferior y superior.

**Salidas:**
1. Matriz triangular inferior $L$.
2. Matriz triangular superior $U$.

In [3]:
import scipy as sp
import numpy as np

# Definimos la matriz de coeficientes A
A = np.array([
    [0, 2, -1],
    [1, 1, 4],
    [6, -2, 3]
], dtype=float)

# Obtenemos la matriz triangular superior
L,U = sp.linalg.lu(A, permute_l=True)

# Mostramos el resultado
print("Matriz A:")
print(A)
print("Matriz L:")
print(L)
print("Matriz U:")
print(U)
print("Matriz L * U :")
print(np.dot(L,U) )


Matriz A:
[[ 0.  2. -1.]
 [ 1.  1.  4.]
 [ 6. -2.  3.]]
Matriz L:
[[0.         1.         0.        ]
 [0.16666667 0.66666667 1.        ]
 [1.         0.         0.        ]]
Matriz U:
[[ 6.         -2.          3.        ]
 [ 0.          2.         -1.        ]
 [ 0.          0.          4.16666667]]
Matriz L * U :
[[ 0.  2. -1.]
 [ 1.  1.  4.]
 [ 6. -2.  3.]]


##### **Ejercicio 3:  Intercambio de filas en la matriz**

**Descripción del problema:**

Manipule las líneas de la matriz $A$ dada a fin de obtener una nueva matriz $B$ que tenga las filas de índice 0 y 2 intercambiadas usando el comando B[[0,2]] = A[[2,0]].

**Objetivo:**
- Reordenar las filas de la matriz.

**Salidas:**
1. Matriz $B$ con filas intercambiadas.

In [6]:
import numpy as np

# Definimos la matriz A (ejemplo)
A = np.array([
    [0, 2, -1],
    [1, 1, 4],
    [6, -2, 3]
], dtype=float)

print("Matriz A original:")
print(A)

# Intercambiamos las filas 0 y 2 usando B[[0,2]] = A[[2,0]]
B = A.copy()  # Creamos una copia para no modificar A
B[[0,2]] = B[[2,0]]  # Intercambia las filas 0 y 2

print("\nMatriz B (filas 0 y 2 intercambiadas):")
print(B)

Matriz A original:
[[ 0.  2. -1.]
 [ 1.  1.  4.]
 [ 6. -2.  3.]]

Matriz B (filas 0 y 2 intercambiadas):
[[ 6. -2.  3.]
 [ 1.  1.  4.]
 [ 0.  2. -1.]]


##### **Ejercicio 4 (difícil): Transformación a matriz estrictamente diagonalmente dominante**

**Descripción del problema:**

Determine una nueva matriz $B$ a partir de $A$ que sea estrictamente diagonalmente dominante mediante el reordenamiento de las filas. Una matriz es estrictamente diagonalmente dominante si, en cada fila, el valor absoluto del elemento en la diagonal principal es mayor que la suma de los valores absolutos de todos los demás elementos de esa misma fila. **Tarea adicional**: piense e implemente una estrategia para realizar este ordenamiento para cualquier matriz $A$.

**Objetivo:**
- Reorganizar las filas de la matriz para hacerla estrictamente diagonalmente dominante.

**Salidas:**
1. Matriz $B$ reordenada.

In [7]:
import numpy as np

def es_diagonalmente_dominante(matriz):
    # Extrae la diagonal principal de la matriz y aplica valor absoluto
    diagonal = np.diag(np.abs(matriz))

    # Calcula la suma de los valores absolutos de cada fila y resta el elemento diagonal
    suma_filas = np.sum(np.abs(matriz), axis=1) - diagonal

    # Verifica si TODOS los elementos diagonales son mayores que las sumas calculadas
    return np.all(diagonal > suma_filas)

def hacer_diagonalmente_dominante(A):
    n = A.shape[0]

    # Verificar si ya es diagonalmente dominante
    if es_diagonalmente_dominante(A):
        return A.copy()

    # Generar todas las permutaciones posibles de filas
    from itertools import permutations
    for perm in permutations(range(n)): # Genera todas las posibles permutaciones de los valores del
                                        # rango de A (Ej: para A 3x3 genera todas las posibles combinaciones
                                        # de los numeros (1,2,3)) e itera en esas combinaciones
        B = A[list(perm)]               # para la iteracion dada ordena las filas como indica la combinacion
        if es_diagonalmente_dominante(B):  # calcula si la cambinacion hallada es diagonalmente dominante
            return B

    print("No se encontró un reordenamiento que haga la matriz diagonalmente dominante")
    return None

# Matriz de ejemplo
A = np.array([
    [0, 2, -1],
    [1, 1, 4],
    [6, -2, 3]
], dtype=float)


print("Matriz A original:")
print(A)

# Intentamos convertirla en diagonalmente dominante
B = hacer_diagonalmente_dominante(A)

if B is not None:
    print("\nMatriz B (diagonalmente dominante):")
    print(B)
    print("\nVerificación:")
    print("¿Es B diagonalmente dominante?", es_diagonalmente_dominante(B))

Matriz A original:
[[ 0.  2. -1.]
 [ 1.  1.  4.]
 [ 6. -2.  3.]]

Matriz B (diagonalmente dominante):
[[ 6. -2.  3.]
 [ 0.  2. -1.]
 [ 1.  1.  4.]]

Verificación:
¿Es B diagonalmente dominante? True


##### **Ejercicio 5: Cálculo de una secuencia definida por una matriz**

**Descripción del problema:**

Calcule los términos de la secuencia definida por $A_n = A^n - 1/2*I$, donde $I$ es la matriz identidad, recibiendo como entrada el valor de $n$.

**Objetivo:**
- Calcular la secuencia definida a partir de la matriz $A$.

**Entradas:**
1. Número natural $n$.

**Salidas:**
1. Matriz resultante $A_n$.

In [ ]:
import numpy as np

def calcular_secuencia_A(A, n):
    error = ""
    if n < 0:
        error = "n debe ser un número natural (n ≥ 0)"

    # Verificar que A sea cuadrada e invertible
    if A.shape[0] != A.shape[1]:
        error = "La matriz A debe ser cuadrada"
    if np.linalg.det(A) == 0:
        error = "La matriz A debe ser invertible"

    I = np.eye(A.shape[0])  # Matriz identidad
    secuencia = []

    for k in range(n + 1):
        A_pow = np.linalg.matrix_power(A, k)  # Calcula A^k
        An = A_pow -((1/2)* I)              # Divide por 2*I (equivale a (1/2)*(A^k)^-1)
        secuencia.append(An)

    return secuencia,error

# Ejemplo de uso

# Definimos la matriz A
A = np.array([
    [0, 2, -1],
    [1, 1, 4],
    [6, -2, 3]
], dtype=float)

n = int(input("Ingrese el valor de n (≥ 0): "))

# Calculamos la secuencia
secuencia,error = calcular_secuencia_A(A, n)

# Mostramos resultados
if error == "":
  k=n
  print(f"\nA{n} =")
  print(np.round(secuencia[n], 4))  # Redondeamos a 4 decimales para mejor visualización
else:
  print(error)

Ingrese el valor de n (≥ 0): 3

A3 =
[[-0.5  2.  -1. ]
 [ 1.   0.5  4. ]
 [ 6.  -2.   2.5]]


#### **Métodos Iterativos: método de Jacobi**

Los "Métodos Iterativos" constituyen una clase de técnicas numéricas para resolver sistemas de ecuaciones lineales, en las cuales la solución exacta $x$ se obtiene como el límite de una secuencia de aproximaciones $x^{(0)}, x^{(1)}, \ldots$. Estos métodos son particularmente útiles cuando las matrices de los sistemas poseen características especiales, como ser grandes y dispersas (es decir, contener muchos elementos nulos).

A diferencia de los métodos directos, que buscan la solución exacta en un número fijo de pasos, los métodos iterativos construyen aproximaciones sucesivas, reduciendo el costo computacional en cada etapa.

En esta sección, estudiaremos el método de Jacobi, una técnica iterativa ampliamente utilizada.


##### **Ejemplo 1: Aplicación del método de Jacobi**



**Descripción del problema:**

Apliquemos el método de Jacobi al siguiente sistema de ecuaciones:
$$
\left\{
\begin{array}{rcc}
4x_1  -x_2  +2x_3 & = & 15\\
  -3x_2  +x_3 & = & 6\\
-x_1  +2x_2  +5x_3 & = & 11
\end{array}
\right.
$$

y calculemos las 4 primeras aproximaciones $x^{(1)}, x^{(2)}, x^{(3)}, x^{(4)}$ tomando como aproximación inicial $x^{(0)}$ el vector nulo.

**Objetivo:**
- Implementar el método de Jacobi iterativamente.
- Comparar las aproximaciones con la solución exacta.

**Salidas:**
1. Aproximaciones $x^{(1)}, x^{(2)}, x^{(3)}, x^{(4)}$.

In [ ]:
import numpy as np

A = np.array([
    [4, -1, 2],
    [0, -3, 1],
    [-1, 2, 5]
], dtype=float)

b = np.array([15, 6, 11], dtype=float)

# Aproximación inicial
x = np.zeros_like(b, dtype=float)

# Número de iteraciones
num_iteraciones = 400

# Matriz diagonal D
D = np.diag(np.diag(A))
D_inv = np.linalg.inv(D)
N = D - A

# Iteraciones de Jacobi
for k in range(num_iteraciones):
    x = np.dot(D_inv, np.dot(N, x) + b)
    print(f"Iteración {k+1}: {x}")
    resultado = np.dot(A, x)
    solucion = True
    for i in range(len(b)):
        if b[i] != resultado[i]:
            solucion = False
    if solucion :
        break

Iteración 1: [ 3.75 -2.    2.2 ]
Iteración 2: [ 2.15       -1.26666667  3.75      ]
Iteración 3: [ 1.55833333 -0.75        3.13666667]
Iteración 4: [ 1.99416667 -0.95444444  2.81166667]
Iteración 5: [ 2.10555556 -1.06277778  2.98061111]
Iteración 6: [ 1.994      -1.00646296  3.04622222]
Iteración 7: [ 1.97527315 -0.98459259  3.00138519]
Iteración 8: [ 2.00315926 -0.99953827  2.98889167]
Iteración 9: [ 2.0056696  -1.00370278  3.00044716]
Iteración 10: [ 1.99885073 -0.99985095  3.00261503]
Iteración 11: [ 1.99872975 -0.99912832  2.99971052]
Iteración 12: [ 2.00036266 -1.00009649  2.99939728]
Iteración 13: [ 2.00027724 -1.00020091  3.00011113]
Iteración 14: [ 1.99989421 -0.99996296  3.00013581]
Iteración 15: [ 1.99994136 -0.99995473  2.99996402]
Iteración 16: [ 2.00002931 -1.00001199  2.99997016]
Iteración 17: [ 2.00001192 -1.00000995  3.00001066]
Iteración 18: [ 1.99999218 -0.99999645  3.00000636]
Iteración 19: [ 1.99999771 -0.99999788  2.99999702]
Iteración 20: [ 2.00000202 -1.00000099 

En la recta real, la distancia entre dos puntos se mide utilizando el valor absoluto. En espacios reales de dimensión mayor, la noción de distancia se generaliza mediante la norma euclidiana, es decir, si $x=(x_1, \ldots, x_n)$ e $y=(y_1, \ldots,y_n) \in \mathbb{R}^n $ entonces

$$||x-y|| = \left(\sum_{i=1}^n(x_i-y_i)^2\right)^{1/2}. $$

Esta medida nos permite evaluar la magnitud de un vector y su proximidad a otros puntos en el espacio. A continuación, mostramos cómo calcular esta norma usando Python.

##### **Ejemplo 2: Cálculo de la norma euclidiana**

**Descripción del problema:**

Calcular la norma euclidiana de un vector en $\mathbb{R}^n$.

**Objetivo:**
- Aplicar la función `np.linalg.norm` para calcular la norma de un vector.

**Salidas:**
1. Valor de la norma euclidiana.

In [ ]:
import numpy as np

# Definimos un vector
x = np.array([3, 4, 5])

# Calculamos la norma euclidiana
norma = np.linalg.norm(x)

print("Norma euclidiana del vector:", norma)

Norma euclidiana del vector: 7.0710678118654755


##### **Ejercicio 1: Implementación del método de Jacobi con criterio de convergencia**

**Descripción del problema:**

Implemente una función en Python que aplique el método de Jacobi para resolver el sistema de ecuaciones
$$
\left\{
\begin{array}{rcc}
5x_1  -2x_2  +3x_3 & = & 8\\
3x_1   +8x_2  -4x_3 & = & 10\\
x_1  -5x_2  +7x_3 & = & 6
\end{array}
\right.
$$
hasta alcanzar una tolerancia de $10^{-4}$.

**Objetivo:**
- Implementar iteraciones hasta la convergencia utilizando la norma euclidiana.

**Salidas:**
1. Vector solución aproximada.



In [8]:
import numpy as np

def metodo_jacobi(A, b, tol=1e-4, max_iter=1000):
    # Resuelve el sistema Ax = b usando el método de Jacobi.
    # Args:
    #    A (np.ndarray): Matriz de coeficientes (n x n)
    #    b (np.ndarray): Vector de términos independientes (n,)
    #    tol (float): Tolerancia para el criterio de parada
    #    max_iter (int): Número máximo de iteraciones permitidas
    #Retorna:
    #    np.ndarray: Vector solución x
    #    int: Número de iteraciones realizadas

    n = len(b)
    x = np.zeros(n)  # Solución inicial (vector cero)
    x_nuevo = np.zeros(n)
    iteraciones = 0
    D = np.diag(np.diag(A))
    D_inv = np.linalg.inv(D)
    N = D - A


    while np.linalg.norm(b-np.dot(A,x)) >= tol:
        x_nuevo = np.dot(D_inv, np.dot(N, x) + b)
        # print(f"Iteración {iteraciones+1}: {x}")
        if iteraciones >= max_iter:
            break


        # Criterio de convergencia (norma euclidiana del cambio relativo)
        error = np.linalg.norm(x_nuevo - x) / np.linalg.norm(x_nuevo)
        if error < tol:
            break

        x = x_nuevo.copy()
        iteraciones += 1

    return x_nuevo, iteraciones

# Sistema de ecuaciones dado
A = np.array([
    [5, -2, 3],
    [3, 8, -4],
    [1, -5, 7]
], dtype=float)

b = np.array([8, 10, 6], dtype=float)

# Resolver usando Jacobi
error=1e-4
solucion, iteraciones = metodo_jacobi(A, b,error,1000)

print(f"Solución aproximada: {solucion}")
print(f"Número de iteraciones: {iteraciones}")
print(f"Verificación (A . solucion): {np.dot(A, solucion)}")


Solución aproximada: [1.13059838 1.82565539 1.99952284]
Número de iteraciones: 21
Verificación (A . solucion): [8.00024963 9.99894687 5.99898132]


##### **Ejercicio 2: Análisis de convergencia del método de Jacobi**

**Descripción del problema:**

No todas las matrices de coeficientes garantizan la convergencia del método de Jacobi hacia la solución exacta del sistema. Use el código desarrollado en el ejercicio anterior para observar los vectores $x^{(0)}, x^{(1)}, \ldots$ generados por el método de Jacobi cuando se aplica a los siguientes:
```python
A = np.array([[1, 3, -2], [2, 6, -4], [-1, -3, 2]], dtype=float)
b = np.array([5, 9, -4], dtype=float)
```


**Objetivo:**
- Analizar si la secuencia generada por el método de Jacobi converge o diverge.
- Reflexionar sobre la preparación del código para manejar situaciones de divergencia.

**Salidas:**

El estudiante debe analizar las iteraciones obtenidas e identificar si la secuencia muestra un comportamiento creciente o errático, indicando falta de convergencia.


In [9]:
import numpy as np

A = np.array([[1, 3, -2], [2, 6, -4], [-1, -3, 2]], dtype=float)
b = np.array([5, 9, -4], dtype=float)

# Aproximación inicial
x = np.zeros_like(b, dtype=float)

# Número de iteraciones
error = 10**-4

# Matriz diagonal D
D = np.diag(np.diag(A))
D_inv = np.linalg.inv(D)
N = D - A
x = np.dot(D_inv, np.dot(N, x) + b)
k = 0

## SOLUCIONES 1.13,1.82,2

# Iteraciones de Jacobi
while np.linalg.norm(b-np.dot(A,x)) >= error:
    x = np.dot(D_inv, np.dot(N, x) + b)
    print(f"Iteración {k+1}: {x}")
    k = k+1
    if k == 1000:
        break

Iteración 1: [-3.5  -1.5   2.75]
Iteración 2: [15.   4.5 -6. ]
Iteración 3: [-20.5   -7.5   12.25]
Iteración 4: [ 52.   16.5 -23.5]
Iteración 5: [-91.5  -31.5   48.75]
Iteración 6: [197.   64.5 -95. ]
Iteración 7: [-378.5  -127.5   193.25]
Iteración 8: [ 774.   256.5 -382.5]
Iteración 9: [-1529.5   -511.5    769.75]
Iteración 10: [ 3079.   1024.5 -1534. ]
Iteración 11: [-6136.5  -2047.5   3074.25]
Iteración 12: [12296.   4096.5 -6141.5]
Iteración 13: [-24567.5   -8191.5   12290.75]
Iteración 14: [ 49161.   16384.5 -24573. ]
Iteración 15: [-98294.5  -32767.5   49155.25]
Iteración 16: [196618.   65536.5 -98300.5]
Iteración 17: [-393205.5  -131071.5   196611.75]
Iteración 18: [ 786443.   262144.5 -393212. ]
Iteración 19: [-1572852.5   -524287.5    786436.25]
Iteración 20: [ 3145740.   1048576.5 -1572859.5]
Iteración 21: [-6291443.5  -2097151.5   3145732.75]
Iteración 22: [12582925.   4194304.5 -6291451. ]
Iteración 23: [-25165810.5   -8388607.5   12582917.25]
Iteración 24: [ 50331662.   1

Resultados de las primeras iteraciones:

    Iteración 1: [-3.5  -1.5   2.75]

    Iteración 2: [15.   4.5 -6. ]

    Iteración 3: [-20.5   -7.5   12.25]

    Iteración 4: [ 52.   16.5 -23.5]

    Iteración 5: [-91.5  -31.5   48.75]

    Iteración 6: [197.   64.5 -95. ]

    Iteración 7: [-378.5  -127.5   193.25]

    Iteración 8: [ 774.   256.5 -382.5]

    Iteración 9: [-1529.5   -511.5    769.75]

    Iteración 10: [ 3079.   1024.5 -1534. ]

Resultado ultimas Iteraciones:

    Iteración 991: [-6.27837075e+298 -2.09279025e+298  3.13918537e+298]

    Iteración 992: [ 1.25567415e+299  4.18558050e+298 -6.27837075e+298]

    Iteración 993: [-2.51134830e+299 -8.37116099e+298  1.25567415e+299]

    Iteración 994: [ 5.0226966e+299  1.6742322e+299 -2.5113483e+299]

    Iteración 995: [-1.00453932e+300 -3.34846440e+299  5.02269660e+299]

    Iteración 996: [ 2.00907864e+300  6.69692879e+299 -1.00453932e+300]

    Iteración 997: [-4.01815728e+300 -1.33938576e+300  2.00907864e+300]

    Iteración 998: [ 8.03631455e+300  2.67877152e+300 -4.01815728e+300]

    Iteración 999: [-1.60726291e+301 -5.35754304e+300  8.03631455e+300]

    Iteración 1000: [ 3.21452582e+301  1.07150861e+301 -1.60726291e+301]

Conclusiones:

    Divergencia evidente: Los valores crecen exponencialmente en lugar de converger

Causas identificadas:

    La matriz es singular (determinante = 0)

    No es diagonal dominante (ninguna fila cumple el criterio)

    Las filas 1 y 2 son linealmente dependientes (Fila 2 = 2 × Fila 1)

Iteración 1: [-3.5  -1.5   2.75]
Iteración 2: [15.   4.5 -6. ]
Iteración 3: [-20.5   -7.5   12.25]
Iteración 4: [ 52.   16.5 -23.5]
Iteración 5: [-91.5  -31.5   48.75]
Iteración 6: [197.   64.5 -95. ]
Iteración 7: [-378.5  -127.5   193.25]
Iteración 8: [ 774.   256.5 -382.5]
Iteración 9: [-1529.5   -511.5    769.75]
Iteración 10: [ 3079.   1024.5 -1534. ]
Iteración 11: [-6136.5  -2047.5   3074.25]
Iteración 12: [12296.   4096.5 -6141.5]
Iteración 13: [-24567.5   -8191.5   12290.75]
Iteración 14: [ 49161.   16384.5 -24573. ]
Iteración 15: [-98294.5  -32767.5   49155.25]
Iteración 16: [196618.   65536.5 -98300.5]
Iteración 17: [-393205.5  -131071.5   196611.75]
Iteración 18: [ 786443.   262144.5 -393212. ]
Iteración 19: [-1572852.5   -524287.5    786436.25]
Iteración 20: [ 3145740.   1048576.5 -1572859.5]
Iteración 21: [-6291443.5  -2097151.5   3145732.75]
Iteración 22: [12582925.   4194304.5 -6291451. ]
Iteración 23: [-25165810.5   -8388607.5   12582917.25]
Iteración 24: [ 50331662.   1

##### **Ejercicio 3: Comparación de normas para evaluar la convergencia**


**Descripción del problema:**

El criterio de convergencia del método de Jacobi puede evaluarse utilizando diferentes normas vectoriales. En este ejercicio, se comparará el desempeño del método utilizando la norma euclidiana (L2), la norma del máximo (L∞) y la norma uno (L1).

**Objetivo:**
- Implementar el método de Jacobi con diferentes normas de convergencia.
- Comparar el número de iteraciones necesarias según la norma utilizada.
- Reflexionar sobre qué norma proporciona una evaluación más eficiente de la convergencia.

**Salidas:**
1. Número de iteraciones necesarias para alcanzar la tolerancia bajo cada norma.
2. Comparación de la velocidad de convergencia.

**Nota:**

El comando `np.linalg.norm` puede utilizar diferentes normas alterando el valor del parámetro `ord` usando un diccionario, por ejemplo:
```python
norm_types = {'L1': 1, 'L2': 2, 'L∞': np.inf}
```
Más detalles se pueden encontrar en la documentación oficial: [numpy.linalg.norm](https://numpy.org/doc/stable/reference/generated/numpy.linalg.norm.html).


In [10]:
import numpy as np

A = np.array([[1, 3, -2], [2, 6, -4], [-1, -3, 2]], dtype=float)
b = np.array([5, 9, -4], dtype=float)
norm_types = {'L1': 1, 'L2': 2, 'L∞': np.inf}

# Aproximación inicial
x = np.zeros_like(b, dtype=float)

# Número de iteraciones
error = 10**-4

# Matriz diagonal D
D = np.diag(np.diag(A))
D_inv = np.linalg.inv(D)
N = D - A
x = np.dot(D_inv, np.dot(N, x) + b)
k = 0

## SOLUCIONES 1.13,1.82,2

k = 0
x = np.zeros_like(b, dtype=float)
# Iteraciones de Jacobi
while np.linalg.norm(b-np.dot(A,x),norm_types['L1']) >= error:
    x = np.dot(D_inv, np.dot(N, x) + b)
 #   print(f"Iteración {k+1}: {x}")
    k = k+1
    if k == 1000:
        break
print(x,k,np.linalg.norm(b-np.dot(A,x),norm_types['L1']))

k = 0
x = np.zeros_like(b, dtype=float)
# Iteraciones de Jacobi
while np.linalg.norm(b-np.dot(A,x),norm_types['L2']) >= error:
    x = np.dot(D_inv, np.dot(N, x) + b)
#    print(f"Iteración {k+1}: {x}")
    k = k+1
    if k == 1000:
        break
print(x,k,np.linalg.norm(b-np.dot(A,x),norm_types['L2']))

k = 0
x = np.zeros_like(b, dtype=float)
# Iteraciones de Jacobi
while np.linalg.norm(b-np.dot(A,x),norm_types['L∞']) >= error:
    x = np.dot(D_inv, np.dot(N, x) + b)
#   print(f"Iteración {k+1}: {x}")
    k = k+1
    if k == 1000:
        break
print(x,k,np.linalg.norm(b-np.dot(A,x),norm_types['L∞']))

[-1.60726291e+301 -5.35754304e+300  8.03631455e+300] 1000 1.928715492935281e+302
[-1.60726291e+301 -5.35754304e+300  8.03631455e+300] 1000 inf
[-1.60726291e+301 -5.35754304e+300  8.03631455e+300] 1000 9.643577464676406e+301


#### **Métodos Iterativos: método de Gauss-Seidel**

A diferencia del método de Jacobi, el método de Gauss-Seidel utiliza los valores ya calculados de las componentes de la solución en la misma iteración para acelerar la convergencia.

##### **Ejemplo: Aplicación del método de Gauss-Seidel**

**Descripción del problema:**

Resolver el siguiente sistema de ecuaciones:
$$
\left\{
\begin{array}{rcc}
4x_1  -x_2  +2x_3 & = & 15\\
  -3x_2  +x_3 & = & 6\\
-x_1  +2x_2  +5x_3 & = & 11
\end{array}
\right.
$$

**Objetivo:**
- Implementar el método de Gauss-Seidel.
- Obtener aproximaciones sucesivas de la solución.

**Salidas:**
1. Aproximaciones de la solución en cada iteración.

In [11]:
import numpy as np

def gauss_seidel(A, b, x0, max_iter=10, tol=1e-4):
    """
    Implementa el método de Gauss-Seidel para resolver sistemas lineales Ax = b.

    Parámetros:
    A (ndarray): Matriz de coeficientes (n x n).
    b (ndarray): Vector de términos independientes (n).
    x0 (ndarray): Vector inicial para las iteraciones (n).
    max_iter (int): Número máximo de iteraciones.
    tol (float): Tolerancia para el criterio de convergencia.

    Retorna:
    ndarray: Vector solución aproximada.
    int: Número de iteraciones realizadas.
    """
    n = len(b)  # Número de ecuaciones
    x = x0.copy()  # Inicializamos el vector solución con la aproximación inicial

    for k in range(max_iter):
        x_new = np.copy(x)  # Copia del vector solución para actualizar en la iteración

        for i in range(n):
            # Suma de los términos antes de la diagonal principal
            sum1 = np.dot(A[i, :i], x_new[:i])

            # Suma de los términos después de la diagonal principal
            sum2 = np.dot(A[i, i+1:], x[i+1:])

            # Actualizamos el valor de x[i] usando la fórmula del método de Gauss-Seidel
            x_new[i] = (b[i] - sum1 - sum2) / A[i, i]

        # Verificamos la convergencia usando la norma infinito
        if np.linalg.norm(x_new - x, ord=np.inf) < tol:
            return x_new, k + 1

        x = x_new  # Actualizamos el vector solución para la siguiente iteración
        print(f"Solucion encontrada: {x}, Iteracion : {k}")

    return x, max_iter  # Retornamos la última solución y el número máximo de iteraciones

# Definimos la matriz de coeficientes A y el vector de términos independientes b
A = np.array([[4, -1, 2], [0, -3, 1], [-1, 2, 5]], dtype=float)
b = np.array([15, 6, 11], dtype=float)

# Vector inicial (x0) con todos los valores en cero
x0 = np.zeros_like(b)

# Llamamos a la función para resolver el sistema
sol, iteraciones = gauss_seidel(A, b, x0)

# Mostramos la solución aproximada y el número de iteraciones realizadas
print(f"Solución aproximada: {sol}")
print(f"Iteraciones: {iteraciones}")

Solucion encontrada: [ 3.75 -2.    3.75], Iteracion : 0
Solucion encontrada: [ 1.375 -0.75   2.775], Iteracion : 1
Solucion encontrada: [ 2.175 -1.075  3.065], Iteracion : 2
Solucion encontrada: [ 1.94875    -0.97833333  2.98108333], Iteracion : 3
Solucion encontrada: [ 2.014875   -1.00630556  3.00549722], Iteracion : 4
Solucion encontrada: [ 1.995675   -0.99816759  2.99840204], Iteracion : 5
Solucion encontrada: [ 2.00125708 -1.00053265  3.00046448], Iteracion : 6
Solucion encontrada: [ 1.9996346  -0.99984517  2.99986499], Iteracion : 7
Solucion encontrada: [ 2.00010621 -1.000045    3.00003924], Iteracion : 8
Solucion encontrada: [ 1.99996913 -0.99998692  2.99998859], Iteracion : 9
Solución aproximada: [ 1.99996913 -0.99998692  2.99998859]
Iteraciones: 10


##### **Ejercicio 1: Implementación del método de Gauss-Seidel**

**Descripción del problema:**

Implemente el método de Gauss-Seidel y aplíquelo a un sistema de ecuaciones de su elección.

**Objetivo:**
- Comprender la implementación del método.
- Experimentar con diferentes matrices y condiciones iniciales.

**Salidas:**
1. Solución aproximada.
2. Número de iteraciones necesarias.

In [12]:
import numpy as np

def gauss_seidel(A, b, x0, tol, max_iter, norm='L2'):
    x = x0.copy()
    for i in range(max_iter):
        x_new = np.zeros_like(x)
        for j in range(A.shape[0]):
            x_new[j] = (b[j] - np.dot(A[j, :j], x_new[:j]) - np.dot(A[j, j + 1:], x[j + 1:])) / A[j, j]
        if np.linalg.norm((x - x_new),norm_types[norm]) < tol:
            return x_new, i
        x = x_new
    return x, max_iter

# Definimos la matriz de coeficientes A y el vector de términos independientes b
A = np.array([[4, -1, 2], [0, -3, 1], [-1, 2, 5]], dtype=float)
b = np.array([15, 6, 11], dtype=float)

# Vector inicial (x0) con todos los valores en cero
error = 10^-4
norm_types = {'L1': 1, 'L2': 2, 'L∞': np.inf}
k = 0
x = np.zeros_like(b, dtype=float)



print('===================================================================')
print('===========  GAUSS - SIEDEL========================================')
for norm in norm_types:
    print(norm)
    x0 = np.zeros_like(b)
    sol, iteraciones = gauss_seidel(A, b, x0, 10 ** -4, 1000, norm)
    print(f"Solución aproximada: {sol}")
    print(f"Iteraciones: {iteraciones}")
    print(f"Error: {np.linalg.norm(b - np.dot(A, sol), norm_types[norm])},Norma: {norm_types[norm]}")


===========  GAUSS - SIEDEL========================================
L1
Solución aproximada: [ 2.00000897 -1.0000038   3.00000332]
Iteraciones: 10
Error: 6.105207244821997e-05,Norma: 1
L2
Solución aproximada: [ 2.00000897 -1.0000038   3.00000332]
Iteraciones: 10
Error: 4.861236409243127e-05,Norma: 2
L∞
Solución aproximada: [ 2.00000897 -1.0000038   3.00000332]
Iteraciones: 10
Error: 4.6329266393030366e-05,Norma: inf


##### **Ejercicio 2: Comparación con el método de Jacobi**

**Descripción del problema:**

Compare la eficiencia del método de Gauss-Seidel con el de Jacobi en términos de iteraciones necesarias y precisión.

**Objetivo:**
- Implementar ambos métodos.
- Evaluar su rendimiento.

**Salidas:**
1. Comparación de iteraciones y errores.

In [13]:
import numpy as np

def gauss_seidel(A, b, x0, tol, max_iter, norm='L2'):
    x = x0.copy()
    for i in range(max_iter):
        x_new = np.zeros_like(x)
        for j in range(A.shape[0]):
            x_new[j] = (b[j] - np.dot(A[j, :j], x_new[:j]) - np.dot(A[j, j + 1:], x[j + 1:])) / A[j, j]
        if np.linalg.norm((x - x_new),norm_types[norm]) < tol:
            return x_new, i
        x = x_new
    return x, max_iter

# Definimos la matriz de coeficientes A y el vector de términos independientes b
A = np.array([[4, -1, 2], [0, -3, 1], [-1, 2, 5]], dtype=float)
b = np.array([15, 6, 11], dtype=float)

# Vector inicial (x0) con todos los valores en cero
error = 10^-4
norm_types = {'L1': 1, 'L2': 2, 'L∞': np.inf}
k = 0
x = np.zeros_like(b, dtype=float)

D = np.diag(np.diag(A))
D_inv = np.linalg.inv(D)
N = D - A
x = np.dot(D_inv, np.dot(N, x) + b)


# Iteraciones de Jacobi
print('===================================================================')
print('===========  JACOBBI  =============================================')
for norm in norm_types:
    print(norm)
    x = np.zeros_like(b)
    while np.linalg.norm(b-np.dot(A,x),norm_types[norm]) >= error:
        x = np.dot(D_inv, np.dot(N, x) + b)
        k = k+1
        if k >= 1000:
            break
    print(f"Solución aproximada: {x}")
    print(f"Iteraciones: {k}")
    print(f"Error: {np.linalg.norm(b - np.dot(A, x), norm_types[norm])}, Norma: {norm_types[norm]}")

A = np.array([[4, -1, 2], [0, -3, 1], [-1, 2, 5]], dtype=float)
b = np.array([15, 6, 11], dtype=float)
k = 0
x = np.zeros_like(b, dtype=float)
norm_types = {'L1': 1, 'L2': 2, 'L∞': np.inf}
# gauss_seidel

print('===================================================================')
print('===========  GAUSS - SIEDEL========================================')
for norm in norm_types:
    print(norm)
    x0 = np.zeros_like(b)
    sol, iteraciones = gauss_seidel(A, b, x0, 10 ** -4, 1000, norm)
    print(f"Solución aproximada: {sol}")
    print(f"Iteraciones: {iteraciones}")
    print(f"Error: {np.linalg.norm(b - np.dot(A, sol), norm_types[norm])},Norma: {norm_types[norm]}")

===========  JACOBBI  =============================================
L1
Solución aproximada: [ 2. -1.  3.]
Iteraciones: 1000
Error: 0.0, Norma: 1
L2
Solución aproximada: [ 3.75 -2.    2.2 ]
Iteraciones: 1001
Error: 10.288950383785508, Norma: 2
L∞
Solución aproximada: [ 3.75 -2.    2.2 ]
Iteraciones: 1002
Error: 7.749999999999999, Norma: inf
===========  GAUSS - SIEDEL========================================
L1
Solución aproximada: [ 2.00000897 -1.0000038   3.00000332]
Iteraciones: 10
Error: 6.105207244821997e-05,Norma: 1
L2
Solución aproximada: [ 2.00000897 -1.0000038   3.00000332]
Iteraciones: 10
Error: 4.861236409243127e-05,Norma: 2
L∞
Solución aproximada: [ 2.00000897 -1.0000038   3.00000332]
Iteraciones: 10
Error: 4.6329266393030366e-05,Norma: inf


##### **Ejercicio 3 (difícil): Condición de convergencia**

**Descripción del problema:**

Verifique si una matriz dada cumple con el criterio de convergencia para el método de Gauss-Seidel. Aplique el método de Gauss-Seidel al sistema y compare los resultados obtenidos.

**Objetivo:**
- Verificar si la matriz de coeficientes es estrictamente diagonalmente dominante.
- Determinar la convergencia del método de Gauss-Seidel en un caso práctico.

**Salidas:**
1. Mensaje indicando si la matriz cumple el criterio de convergencia.
2. Solución aproximada y número de iteraciones.

In [14]:
import numpy as np

def gauss_seidel(A, b, x0, tol, max_iter):
    x = x0.copy()
    for i in range(max_iter):
        x_new = np.zeros_like(x)
        for j in range(A.shape[0]):
            x_new[j] = (b[j] - np.dot(A[j, :j], x_new[:j]) - np.dot(A[j, j+1:], x[j+1:])) / A[j, j]
        if np.linalg.norm(x_new - x, np.inf) < tol:
            return x_new, i+1
        x = x_new
    return x, max_iter

def diagonalmente_dominante(A):
    diagonal = np.abs(np.diag(A))
    suma_filas = np.sum(np.abs(A), axis=1) - diagonal
    return np.all(diagonal > suma_filas)

def jacobi(A, b, x0, tol, max_iter):
    x = x0.copy()
    for i in range(max_iter):
        x_new = np.zeros_like(x)
        for j in range(A.shape[0]):
            x_new[j] = (b[j] - np.dot(A[j, :j], x[:j]) - np.dot(A[j, j+1:], x[j+1:])) / A[j, j]
        if np.linalg.norm(x_new - x, np.inf) < tol:
            return x_new, i+1
        x = x_new
    return x, max_iter

# Definimos la matriz y vector
A = np.array([[4, -1, 2], [0, -3, 1], [-1, 2, 5]], dtype=float)
b = np.array([15, 6, 11], dtype=float)
x0 = np.zeros_like(b)

# Verificamos condiciones
if diagonalmente_dominante(A):
    print("La matriz ES diagonalmente dominante")
else:
    print("La matriz NO ES diagonalmente dominante")

if np.linalg.det(A) != 0:
    print("La matriz ES invertible")
else:
    print("La matriz NO ES invertible")

if diagonalmente_dominante(A) and np.linalg.det(A) != 0:
    print("\nCumple criterios de convergencia para métodos iterativos")

    # Resolvemos con Gauss-Seidel
    sol_gs, iter_gs = gauss_seidel(A, b, x0, 1e-4, 1000)
    print("\nMétodo de Gauss-Seidel:")
    print(f"Solución aproximada: {sol_gs}")
    print(f"Iteraciones: {iter_gs}")
    print(f"Error (norma infinito): {np.linalg.norm(b - A @ sol_gs, np.inf)}")

    # Resolvemos con Jacobi
    sol_j, iter_j = jacobi(A, b, x0, 1e-4, 1000)
    print("\nMétodo de Jacobi:")
    print(f"Solución aproximada: {sol_j}")
    print(f"Iteraciones: {iter_j}")
    print(f"Error (norma infinito): {np.linalg.norm(b - A @ sol_j, np.inf)}")
else:
    print("\nNo se garantiza convergencia para métodos iterativos")


La matriz ES diagonalmente dominante
La matriz ES invertible

Cumple criterios de convergencia para métodos iterativos

Método de Gauss-Seidel:
Solución aproximada: [ 2.00000897 -1.0000038   3.00000332]
Iteraciones: 11
Error (norma infinito): 4.6329266393030366e-05

Método de Jacobi:
Solución aproximada: [ 2.00002931 -1.00001199  2.99997016]
Iteraciones: 16
Error (norma infinito): 0.00020247340925649837
